### PR TITLE
detect mocked module on saving pass

### PR DESCRIPTION
**Pull request by PaliC**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69793

Raises a not implemented error if we attempt to pickle an object which uses a mocked module. Now we no longer have to load the object to get this check, and instead happens right on the saving path.

Differential Revision: [D33146274](https://our.internmc.facebook.com/intern/diff/D33146274)

### CODE REVIEW

File: torch/package/package_exporter.py [hunk 1]
```diff
@@ -39,7 +39,6 @@
 
 ActionHook = Callable[["PackageExporter", str], None]
 
-
```
Comment (suo) [inline torch/package/package_exporter.py:42]: nit: try not to include unrelated changes like whitespace and formatting in your PR to keep the project history clean.

File: torch/package/package_exporter.py [hunk 2]
```diff
@@ -550,7 +549,10 @@ def _intern_module(
                 self.add_dependency(dep)
 
     def save_pickle(
-        self, package: str, resource: str, obj: Any, dependencies: bool = True
+        self, package: str,
+        resource: str, obj: Any,
+        dependencies: bool = True,
+        save_mocked_and_used: bool = False
```
Comment (suo) [inline torch/package/package_exporter.py:555]: I think we don't need to make this an option for users. While this change is technically BC-incompatible, any package that trips this behavior would not have been importable anyway, so it's probably fine.

**Review by suo — CHANGES_REQUESTED** [review]
Looks pretty good overall, only a small comment about not presenting it as an option in the API. We should make sure internal tests pass on this as well

File: torch/package/package_exporter.py (hunk 1 above)
Comment (PaliC) [reply torch/package/package_exporter.py:42]: ah gotcha, so just lint right before merging then

File: torch/package/package_exporter.py (hunk 2 above)
Comment (PaliC) [reply torch/package/package_exporter.py:555]: Yeah it was primarily for the unit test, I couldn't think of a better way to do the bc test. I'll just delete it then.

**Comment by suo** [comment]
Please import these changes internally so that we can run fbcode unit tests against them

**Comment by PaliC** [comment]
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

**Review by suo — APPROVED** [review]
lgtm, but make sure we don't break anything internal :)

**Comment by PaliC** [comment]
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

**Comment by PaliC** [comment]
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

**Comment by PaliC** [comment]
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

**Comment by PaliC** [comment]
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

**Comment by PaliC** [comment]
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

**Comment by PaliC** [comment]
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

**Comment by PaliC** [comment]
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

**Comment by PaliC** [comment]
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

**Review by suo — APPROVED** [review]

**Comment by PaliC** [comment]
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

**Comment by PaliC** [comment]
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

**Comment by PaliC** [comment]
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).

**Comment by PaliC** [comment]
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33146274).
